mmf/models/transformers/base.py

```diff
@@ -0,0 +1,204 @@
+
+        Example ::
+
+            # For pretraining
+            self.classifier = nn.Sequential(
+                BertPredictionHeadTransform(self.transformer_config),
+                nn.Linear(self.transformer_config.hidden_size, self.config.num_labels),
+            )
+
+            # For classification
+            self.classifier = nn.Sequential(
+                BertPredictionHeadTransform(self.transformer_config),
+                nn.Linear(self.transformer_config.hidden_size, self.config.num_labels),
+            )
+        """
+        pass
+
+    def build_losses(self):
+        """Initialize the losses for pretraining. For example MLM, MIM etc.
+
+        Example ::
+
+            self.mlm_loss = CrossEntropyLoss(ignore_index=-1)
+        """
+        pass
+
+    def _init_weights(self, module):
+        """Initialize the weights.
+        """
+        if isinstance(module, (nn.Linear, nn.Embedding)):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+        elif isinstance(module, nn.module.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        if isinstance(module, nn.Linear) and module.bias is not None:
+            module.bias.data.zero_()
+
+    def tie_weights(self):
+        """Tie the weights between the input embeddings and the output embeddings
+        if required.
+        """
+        pass
+
+    def init_weights(self):
+        if self.config.random_initialize is False:
+            if self.config.transformer_base is None:
+                # No pretrained model, init weights
+                self.apply(self._init_weights)
+
+        # Tie weights if required
+        self.tie_weights()
+
+    def preprocess_sample(self, sample_list: Dict[str, Any]):
+        """Preprocess the sample_list. Generate image input ids, image position ids,
+        image mask. Create a attention mask for all the masks for different modalities.
+        """
+        # TODO(vedanuj): Make this method simpler
+        # Image feature(img_input_ids) and position(img_pos_ids) tokens
+        if "image_feature_0" in sample_list:
+            sample_list.img_input_ids = sample_list.image_feature_0
+            sample_list.img_pos_ids = sample_list.image_info_0.bbox
+        else:
+            sample_list.img_input_ids = self.vision_encoder(sample_list.image)
+            sample_list.img_pos_ids = sample_list.img_input_ids.new_tensor(
+                torch.arange(0, sample_list.img_input_ids.size(1), dtype=torch.long)
+                .unsqueeze(0)
+                .expand(
+                    sample_list.img_input_ids.size(0), sample_list.img_input_ids.size(1)
+                )
+                .unsqueeze(-1)
+            )
+
+        # Image mask
+        sample_list.image_mask = torch.ones_like(
+            sample_list.img_input_ids[:, :, 0], dtype=torch.long
+        )
+
+        # Attention mask
+        sample_list.attention_mask = torch.cat(
```

Comment:
Attention mask shouldn't be assigned to sample list and should be calculated on the fly in the implementation.